### ISSUE TITLE
Randomly incorrect results because SubgraphCallable equality check on name is insufficient

### ISSUE THREAD

**Issue by bmerry** [dask/dask#7632]
**What happened**: Since #6666 (in 2.28.0), `SubgraphCallable.__eq__` does not compare the `dsk` field. Depending on how things have been persisted and optimised, two different SubgraphCallables can have the same name (inherited from the Blockwise which generates them) but be semantically different.

This has caused incorrect results from some code we have in production, I believe because Distributed keeps a cache of task functions, and the wrong version is being cached in some cases due to the pickled form of one version being used for the other. So far I haven't been able to create a MCVE that actually produces incorrect results (some optimisation or caching seem to be interfering), but I can demonstrate two SubgraphCallables that compare equal but operate differently.

**What you expected to happen**: I'm not sure how to fix this without re-introducing the problem that #6666 fixed. I suspect part of the problem is that when Blockwise uses SubgraphCallable, it just names the inputs `_0`, `_1` etc, losing the information about which layers those actually refer to. So when `SubgraphCallable.__eq__` checks that the "inputs" are the same, it's only really checking the arity rather than that the inputs are actually the same. Possibly adding that information to a SubgraphCallable (maybe by hashing it and using it in the name, rather than just calling them all "subgraph_callable") would make it unique, but I'm not sure I understand the issue well enough to be sure.

**Minimal Complete Verifiable Example**:

`x` and `y` are equivalent arrays, but `y` has a slightly flattened graph due to persisting one of the inputs. Persisting doesn't change the array name, so `x` and `y` end up with the same name.

```python
#!/usr/bin/env python3

import numpy as np
import dask
import dask.array as da


def main():
    aa = np.arange(5)
    bb = np.array([10, 7, 8, 3, 5])

    a = da.from_array(aa)
    b = da.from_array(bb)
    mb = -b
    x = a * mb
    mb = mb.persist()
    y = a * mb
    x, = dask.optimize(x)
    y, = dask.optimize(y)
    key = x.__dask_keys__()[0]
    sgc_x = x.__dask_graph__()[key][0]
    sgc_y = y.__dask_graph__()[key][0]
    print(sgc_x.__reduce__())
    print(sgc_y.__reduce__())
    print(sgc_x == sgc_y)

if __name__ == '__main__':
    main()
```

Output:
```
(<class 'dask.optimization.SubgraphCallable'>, ({'mul-b50ea77ca04025da7417b3a2b77c5b9d': 'neg-mul-b50ea77ca04025da7417b3a2b77c5b9d', 'neg-mul-b50ea77ca04025da7417b3a2b77c5b9d': (<built-in function mul>, '_0', (<built-in function neg>, '_1'))}, 'mul-b50ea77ca04025da7417b3a2b77c5b9d', ('_0', '_1'), 'subgraph_callable'))
(<class 'dask.optimization.SubgraphCallable'>, ({'mul-b50ea77ca04025da7417b3a2b77c5b9d': (<built-in function mul>, '_0', '_1')}, 'mul-b50ea77ca04025da7417b3a2b77c5b9d', ('_0', '_1'), 'subgraph_callable'))
True
```

I've printed the SubgraphCallable's via the `__reduce__` method, since `__repr__` hides a lot of the details. The SubgraphCallable for x returns mul(_0, neg(_1)), while that for y returns mul(_0, _1), which are clearly different. Yet they compare equal.

**Environment**:

- Dask version: 2.28, latest main (23c93d7)
- Python version: 3.6, 3.8
- Operating System: Ubuntu 18.04, Ubuntu 20.04
- Install method (conda, pip, source):

**Comment by bmerry** [comment]
I've managed to create an example where the wrong values are produced. It seems to be random whether the bug manifests, so if you don't reproduce it immediately, try running it a few times (I haven't rigorously measured failure rates, but for me it's closer to 50% than 1%).

```
#!/usr/bin/env python3

import numpy as np
import dask
import dask.array as da
import dask.distributed


def main():
    client = dask.distributed.Client()
    aa = np.arange(3)
    bb = np.array([10 + 2j, 7 - 3j, 8 + 1j])

    a = da.from_array(aa)
    b = da.from_array(bb)
    cb = b.conj()
    x = a * cb
    cb, = dask.optimize(cb)
    y = a * cb

    with dask.config.set(optimization__fuse__active=False):
        print(x.compute())
        print(y.compute())

if __name__ == '__main__':
    main()
```

`x` and `y` should be semantically the same but one has an optimised graph. However, output is
```
[ 0.+0.j  7.+3.j 16.-2.j]
[ 0.+0.j  7.-3.j 16.+2.j]
```

Verified with dask 2021.4.1+19.g23c93d72 and distributed 2021.4.1+15.gc76ed460.

**Comment by bmerry** [comment]
This problem is presumably rare (I'm assuming it hasn't regularly been encountered since it's been around for a while) but also high-severity (randomly corrupting calculations). Here's a suggestion for a quick fix: if no name is explicitly given to the SubgraphCallable constructor, it should use `tokenize` on the other arguments to form a unique name, rather than defaulting to `subgraph_callable`. That could be expensive if there is a large numpy array embedded in the subgraph, but at least it would be correct. An alternative would be to generate a UUID. I don't know much about dask.distributed, but it looks like `Blockwise.__dask_distributed_pack__` constructs a re-instantiates the SubgraphCallable each time it is used, so possibly a UUID would prevent reuse of the pickle cache in that case.

@madsbk would tokenizing the subgraph be acceptable for the use case that led to #6666?

@TomAugspurger I see dask.dataframe has one use of `SubgraphCallable, but it passes an explicit name (based on tokenization). Does this name uniquely identify the dsk in the SubgraphCallable?

**Comment by bmerry** [comment]
I've added a PR with my proposed fix (#7637).
